File: cpp/algorithms/quick.cpp
#include "quick.hpp"
#include <vector>
#include <utility> // for std::swap

namespace
{
    int partition(std::vector<int>& numbers, int left_index, int right_index)
    {
        int pivot_value = numbers[right_index];
        int i = left_index - 1;

        for (int j = left_index; j < right_index; ++j)
        {
            if (numbers[j] <= pivot_value)
            {
                ++i;
                std::swap(numbers[i], numbers[j]);
            }
        }
        std::swap(numbers[i + 1], numbers[right_index]);
        return i + 1;
    }

    void quick_sort_recursive(std::vector<int>& numbers, int left_index, int right_index)
    {
        if (left_index >= right_index)
            return;

        int pivot_position = partition(numbers, left_index, right_index);
        quick_sort_recursive(numbers, left_index, pivot_position - 1);
        quick_sort_recursive(numbers, pivot_position + 1, right_index);
    }
}

void quick_sort(std::vector<int>& numbers)
{
    if (numbers.size() < 2)
        return;

    quick_sort_recursive(numbers, 0, numbers.size() - 1);
}

File: python/algorithms/quick.py
def quick_sort(numbers):
    """In-place Quick Sort using Lomuto partition scheme."""
    if not numbers or len(numbers) < 2:
        return

    def partition(nums, left_index, right_index):
        pivot_value = nums[right_index]
        i = left_index - 1
        for j in range(left_index, right_index):
            if nums[j] <= pivot_value:
                i += 1
                nums[i], nums[j] = nums[j], nums[i]
        nums[i + 1], nums[right_index] = nums[right_index], nums[i + 1]
        return i + 1

    def quick_sort_recursive(nums, left_index, right_index):
        if left_index >= right_index:
            return
        pivot_position = partition(nums, left_index, right_index)
        quick_sort_recursive(nums, left_index, pivot_position - 1)
        quick_sort_recursive(nums, pivot_position + 1, right_index)

    quick_sort_recursive(numbers, 0, len(numbers) - 1)
